File: packages/core/src/engine/oldAgePension.ts
import type { YearData, OldAgePensionInput, OldAgePensionResult } from '../types';
import { parseRate, roundHalfUp } from '../rounding/strategies';

function assertNonNeg(name: string, v: number): void {
  if (!Number.isFinite(v) || v < 0) throw new Error(`${name} must be a finite non-negative number, got ${v}`);
}

export function calcOldAgePension(data: YearData, input: OldAgePensionInput): OldAgePensionResult {
  const oap = data.oldAgePension;
  if (!oap) throw new Error(`No oldAgePension data for year ${data.year}（該年度無老年年金資料）`);
  const months = input.months ?? 0;
  const offset = input.claimOffsetMonths ?? 0;
  assertNonNeg('avgInsuredSalary', input.avgInsuredSalary);
  assertNonNeg('years', input.years);
  assertNonNeg('months', months);
  if (months > 11 || !Number.isInteger(months)) throw new Error(`months must be an integer 0–11, got ${months}`);
  if (!Number.isInteger(input.years)) throw new Error(`years must be an integer, got ${input.years}`);
  if (!Number.isInteger(offset)) throw new Error(`claimOffsetMonths must be an integer, got ${offset}`);

  const M = input.years * 12 + months;
  const cap = oap.maxAdjustYears * 12;
  const adjMonths = Math.max(-cap, Math.min(cap, offset));
  const a = parseRate(oap.adjustPerYearRate); // {num:4, den:100}
  const factorN = 12 * a.den + adjMonths * a.num;
  const factorD = 12 * a.den;

  const ra = parseRate(oap.formulaARate); // {num:775, den:100000}
  const formulaA = roundHalfUp(
    (input.avgInsuredSalary * M * ra.num + oap.formulaABonus * 12 * ra.den) * factorN,
    12 * ra.den * factorD,
  );
  const rb = parseRate(oap.formulaBRate); // {num:155, den:10000}
  const formulaB = roundHalfUp(input.avgInsuredSalary * M * rb.num * factorN, 12 * rb.den * factorD);

  return {
    formulaA,
    formulaB,
    monthly: Math.max(formulaA, formulaB),
    adjustmentMonths: adjMonths,
    eligible: M >= oap.minYearsForPension * 12,
  };
}

export function averageHighestInsuredSalary(monthlySalaries: number[]): number {
  if (monthlySalaries.length === 0) throw new Error('monthlySalaries must not be empty');
  monthlySalaries.forEach((s, i) => assertNonNeg(`monthlySalaries[${i}]`, s));
  const top = [...monthlySalaries].sort((x, y) => y - x).slice(0, 60);
  return roundHalfUp(
    top.reduce((s, v) => s + v, 0),
    top.length,
  );
}

export function statutoryClaimAge(data: YearData, bornRocYear: number): number {
  const sa = data.oldAgePension?.statutoryAge;
  if (!sa) throw new Error(`No oldAgePension data for year ${data.year}`);
  const hit = sa.schedule.find((e) => bornRocYear <= e.maxBornRocYear);
  return hit ? hit.age : sa.defaultAge;
}

File: packages/python/taiwan_payroll/engine/old_age_pension.py
from __future__ import annotations

from .._rounding import parse_rate, round_half_up, not_finite_number
from .._types import OldAgePensionInput, OldAgePensionResult


def _assert_non_neg(name: str, v: float) -> None:
    if not_finite_number(v) or v < 0:
        raise ValueError(f"{name} must be a finite non-negative number, got {v}")


def calc_old_age_pension(data: dict, inp: OldAgePensionInput) -> OldAgePensionResult:
    oap = data.get("oldAgePension")
    if not oap:
        raise ValueError(f"No oldAgePension data for year {data.get('year')}（該年度無老年年金資料）")
    months = inp.months
    offset = inp.claim_offset_months
    _assert_non_neg("avg_insured_salary", inp.avg_insured_salary)
    _assert_non_neg("years", inp.years)
    _assert_non_neg("months", months)
    if not (isinstance(months, int) and 0 <= months <= 11):
        raise ValueError(f"months must be an integer 0–11, got {months}")
    if not isinstance(inp.years, int) or isinstance(inp.years, bool):
        raise ValueError(f"years must be an integer, got {inp.years}")
    if not isinstance(offset, int) or isinstance(offset, bool):
        raise ValueError(f"claim_offset_months must be an integer, got {offset}")

    M = int(inp.years) * 12 + months
    cap = oap["maxAdjustYears"] * 12
    adj = max(-cap, min(cap, offset))
    a_num, a_den = parse_rate(oap["adjustPerYearRate"])
    fN = 12 * a_den + adj * a_num
    fD = 12 * a_den
    an, ad = parse_rate(oap["formulaARate"])
    formula_a = round_half_up(
        (int(inp.avg_insured_salary) * M * an + oap["formulaABonus"] * 12 * ad) * fN,
        12 * ad * fD,
    )
    bn, bd = parse_rate(oap["formulaBRate"])
    formula_b = round_half_up(int(inp.avg_insured_salary) * M * bn * fN, 12 * bd * fD)
    return OldAgePensionResult(
        formula_a=formula_a,
        formula_b=formula_b,
        monthly=max(formula_a, formula_b),
        adjustment_months=adj,
        eligible=M >= oap["minYearsForPension"] * 12,
    )


def average_highest_insured_salary(monthly_salaries: list) -> int:
    if not monthly_salaries:
        raise ValueError("monthly_salaries must not be empty")
    for i, s in enumerate(monthly_salaries):
        _assert_non_neg(f"monthly_salaries[{i}]", s)
    top = sorted(monthly_salaries, reverse=True)[:60]
    return round_half_up(sum(int(x) for x in top), len(top))


def statutory_claim_age(data: dict, born_roc_year: int) -> int:
    sa = (data.get("oldAgePension") or {}).get("statutoryAge")
    if not sa:
        raise ValueError(f"No oldAgePension data for year {data.get('year')}")
    for e in sa["schedule"]:
        if born_roc_year <= e["maxBornRocYear"]:
            return e["age"]
    return sa["defaultAge"]
